File: 45-jump-game-ii/45-jump-game-ii.cpp
class Solution {
public:
    int jump(vector<int>& nums) {
        int n=nums.size();
            if(n==1) return 0;
            
            int currReach= 0; 
            int maxReach= 0;
            int minJumps=0;
            
            for(int i=0;i<n-1;i++){
                    maxReach=max(i+nums[i],maxReach);
                    
                    if(i==currReach){
                            minJumps++;
                            currReach=maxReach;
                    }
            }
      return minJumps;
    }
};

File: 45-jump-game-ii/45-jump-game-ii.java
class Solution {
    public int jump(int[] nums) {
            //greedy way
        int curReach = 0; // i=0 kaha tk le kr ja skta
        int maxReach = 0; // i=0 ki reach tk aa rhe elements, kha tk max le kr ja skte
            //i=0 has nums[i]=3, so any one of 1,2,3 will give maxReach
        int jumps = 0;
        
        for(int i = 0;i<nums.length-1;i++){
            if(i+nums[i] > maxReach){
                maxReach = i +nums[i];
            }
            
            if(i==curReach){
                jumps++; // i=0 se 1,2,3 me se koi ek that gives maxReach
                curReach = maxReach;
            }
        }
        return jumps;
    }
}